### DISCUSSION TITLE
Foundry fundamentals : Issue in deploying lottery smart contract.

### DISCUSSION THREAD

**Discussion by Rayyan-Uddin** [Cyfrin/foundry-full-course-cu#3116] [Q&A] [answered]
Got this error while writing deployment script tried to debug but can't find anything, help.
![error](https://github.com/user-attachments/assets/0f1217bd-5536-4852-9668-14ce5de9b2a3)

**Comment by EngrPips** [comment]
Hello @Rayyan-Uddin, Please show your `HelperConfig` and your `Raffle` contract. Please format them nicely.

**Reply by Rayyan-Uddin** [reply]
![error1](https://github.com/user-attachments/assets/914c3e65-1b45-43d7-87e3-f3455e69ecaf)
![error2](https://github.com/user-attachments/assets/0856c423-38b1-4bcc-99d6-728a77e84502)

**Reply by Rayyan-Uddin** [reply]
I hope these screenshots are visible, let me know if not. Thanks.

**Reply by EngrPips** [reply]
Please share code instead of picture, copy the code and send it. Make sure to edit it nicely

**Reply by sanu2002** [reply]
send the code

**Reply by Rayyan-Uddin** [reply]
HelperConfig.s.sol


// SPDX-License-Identifier: MIT
pragma solidity 0.8.19;

import {Script} from "forge-std/Script.sol";
import {VRFCoordinatorV2_5Mock} from "@chainlink/contracts/src/v0.8/vrf/mocks/VRFCoordinatorV2_5Mock.sol";

abstract contract CodeConstants {
    // mock VRF Values
    uint96 public MOCK_BASE_FEE = 0.25 ether;
    uint96 public MOCK_GAS_PRICE_LINK = 1e9;
    // LINK / Eth price
    int256 public MOCK_WEI_PER_UNIT_LINK = 4e15;

    uint256 public constant ETH_SEPOLIA_CHAIN_ID = 11155111;
    uint256 public constant LOCAL_CHAIN_ID = 31337;
}
contract HelperConfig is CodeConstants, Script {

    error HelperConfig__InvalidChainId();

    struct NetworkConfig {
        uint256 entranceFee;
        uint256 interval;
        address vrfCoordinator;
        bytes32 gasLane;
        uint32 callbackGasLimit;
        uint256 subscriptionId;
    }
    NetworkConfig public localNetworkConfig;
    mapping(uint256 chainid => NetworkConfig) public networkConfigs;

    constructor() {
        networkConfigs[ETH_SEPOLIA_CHAIN_ID] = getSepoliaEthConfig();
    }

    function getConfigByChainId(uint256 chainId) public returns (NetworkConfig memory) {
        if (networkConfigs[chainId].vrfCoordinator != address(0)) {
            return networkConfigs[chainId];
        } else if (chainId == LOCAL_CHAIN_ID) {
            return getOrCreateAnvilEthConfig();
        } else {
            revert HelperConfig__InvalidChainId();
        }
    }

    function getConfig() public returns (NetworkConfig memory) {
        return getConfigByChainId(block.chainid);
    }

    function getSepoliaEthConfig() public pure returns (NetworkConfig memory) {
        return
            NetworkConfig({
                entranceFee: 0.01 ether,
                interval: 30,
                vrfCoordinator: 0x9DdfaCa8183c41ad55329BdeeD9F6A8d53168B1B,
                gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
                callbackGasLimit: 500000,
                subscriptionId: 0
            });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory) {
        if (localNetworkConfig.vrfCoordinator != address(0)) {
            return localNetworkConfig;
        }
        // Deploy mocks
        vm.startBroadcast();
        VRFCoordinatorV2_5Mock vrfCoordinatorMock = new VRFCoordinatorV2_5Mock(
            MOCK_BASE_FEE,
            MOCK_GAS_PRICE_LINK,
            MOCK_WEI_PER_UNIT_LINK
        );
        vm.stopBroadcast();

        localNetworkConfig = NetworkConfig({
            entranceFee: 0.01 ether,
            interval: 30,
            vrfCoordinator: address(vrfCoordinatorMock),
            gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
            callbackGasLimit: 500000,
            subscriptionId: 0
        });
        return localNetworkConfig;
    }
}

**Reply by Rayyan-Uddin** [reply]
Raffle.sol


pragma solidity ^0.8.19;

import {VRFConsumerBaseV2Plus} from "@chainlink/contracts/src/v0.8/vrf/dev/VRFConsumerBaseV2Plus.sol";
import {VRFV2PlusClient} from "@chainlink/contracts/src/v0.8/vrf/dev/libraries/VRFV2PlusClient.sol";

contract Raffle is VRFConsumerBaseV2Plus {
    //Errors
    error Raffle__sendMoreEth();
    error Raffle__transferFail();
    error Raffle__isNotOpen();
    error Raffle__upkeepnotNeeded(
        uint256 balance,
        uint256 playersLength,
        uint256 raffleState
    );

    // Type declaration
    enum RaffleState {
        OPEN,
        CALCULATING
    }

    //State Variables
    uint16 private constant REQUEST_CONFIRMATION = 3;
    uint32 private constant NUM_WORD = 1;
    uint256 private immutable i_entrancePrice;
    uint256 private immutable i_interval;
    bytes32 private immutable i_keyHash;
    uint256 private immutable i_subscriptionId;
    uint32 private immutable i_callbackGasLimit;

    address payable[] private s_players;
    uint256 private s_lastTimeStamp;
    address payable private s_recentWinner;
    RaffleState private s_raffleState = RaffleState.OPEN; //start as open

    //Events
    event RaffleEntered(address indexed player);

    //Constructor
    constructor(
        uint256 entrancePrice,
        uint256 interval,
        bytes32 gasLane,
        uint256 subscriptionId,
        uint32 callbackGasLimit,
        address vrfCoordinator
    ) VRFConsumerBaseV2Plus(vrfCoordinator) {
        i_entrancePrice = entrancePrice;
        i_keyHash = gasLane;
        i_subscriptionId = subscriptionId;
        i_callbackGasLimit = callbackGasLimit;
        // @dev duration of lottery in seconds
        i_interval = interval;
        s_lastTimeStamp = block.timestamp;
        // s_vrfCoordinator.requestRandomWords();
    }

    //Functions
    function enterRaffle() external payable {
        if (msg.value < i_entrancePrice) {
            revert Raffle__sendMoreEth();
        }
        if (s_raffleState != RaffleState.OPEN) {
            revert Raffle__isNotOpen();
        }
        s_players.push(payable(msg.sender));

        emit RaffleEntered(msg.sender);
    }

function checkUpkeep(
        bytes memory /*checkData*/
    ) public view returns (bool upkeepNeeded, bytes memory /*performData*/) {
        bool timeHasPassed = ((block.timestamp - s_lastTimeStamp) >=
            i_interval);
        bool isOpen = s_raffleState == RaffleState.OPEN;
        bool hasPlayers = s_players.length > 0;
        upkeepNeeded = timeHasPassed && isOpen && hasPlayers;
        return (upkeepNeeded, "");
    }

    // get a random number
    // check if enough time has passed
    function performUpkeep(bytes calldata /* performData */) external {
        //check if enough time passed to pickwinner
        (bool upkeepNeeded, ) = checkUpkeep("");
        if (!upkeepNeeded) {
            revert Raffle__upkeepnotNeeded(
                address(this).balance,
                s_players.length,
                uint256(s_raffleState)
            );
        }
        s_raffleState = RaffleState.CALCULATING;

        VRFV2PlusClient.RandomWordsRequest memory request = VRFV2PlusClient
            .RandomWordsRequest({
                keyHash: i_keyHash,
                subId: i_subscriptionId,
                requestConfirmations: REQUEST_CONFIRMATION,
                callbackGasLimit: i_callbackGasLimit,
                numWords: NUM_WORD,
                extraArgs: VRFV2PlusClient._argsToBytes(
                    // Set nativePayment to true to pay for VRF requests with Sepolia ETH instead of LINK
                    VRFV2PlusClient.ExtraArgsV1({nativePayment: false})
                )
            });

        uint256 requestId = s_vrfCoordinator.requestRandomWords(request);
    }

    function fulfillRandomWords(
        uint256 requestId,
        uint256[] calldata randomWords
    ) internal override {
        uint256 indexOfWinner = randomWords[0] % s_players.length;
        address payable recentWinner = s_players[indexOfWinner];
        s_recentWinner = recentWinner;
        s_raffleState = RaffleState.OPEN;
        (bool success, ) = recentWinner.call{value: address(this).balance}("");
        if (!success) {
            revert Raffle__transferFail();
        }
    }

    /**
     * Getter function
     */
    function getEntrancePrice() external view returns (uint256) {
        return i_entrancePrice;
    }

    function getRaffleState() external view returns (RaffleState) {
        return s_raffleState;
    }
}

**Comment by EngrPips** [comment, marked as answer]
> Got this error while writing deployment script tried to debug but can't find anything, help. ![error](https://private-user-images.githubusercontent.com/110310731/399024262-0f1217bd-5536-4852-9668-14ce5de9b2a3.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU0NTg4NzcsIm5iZiI6MTczNTQ1ODU3NywicGF0aCI6Ii8xMTAzMTA3MzEvMzk5MDI0MjYyLTBmMTIxN2JkLTU1MzYtNDg1Mi05NjY4LTE0Y2U1ZGU5YjJhMy5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyOVQwNzQ5MzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YTE2NTNkYmRlYWEzNDc2ZjRjZDk4MDM3MzllMTM2MDA0Yzc2MWU4NzY4MmJkZThjMDlkMTIzNzY0NmMzZTM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Nt3Bccux-nY8Z1N6gKGzBeFEfzy2cWe_eZmeD2nQSdc)

Hello @Rayyan-Uddin, So the issue here is you are wrongly passing the arguments to the `Raffle` contract constructor. subscriptionId comes as the fifth argument in the `Raffle` contract constructor, but you are passing it as the sixth. Please pass the constructor argument in the correct order so things work correctly.

**Reply by Rayyan-Uddin** [reply]
Yup, thanks.

**Reply by EngrPips** [reply]
Anytime, Friend. Keep crushing it.
